xapian-core/cluster/cluster.cc

```diff
@@ -0,0 +1,586 @@
+	    termfreq[*titer] = 1;
+	else
+	    i->second += 1;
+    }
+}
+
+doccount
+TermListGroup::get_doccount() const
+{
+    LOGCALL(API, doccount, "TermListGroup::get_doccount", NO_ARGS);
+    return docs_num;
+}
+
+doccount
+TermListGroup::get_termfreq(const string &tname) const
+{
+    LOGCALL(API, doccount, "TermListGroup::get_termfreq", tname);
+    unordered_map<string, doccount>::const_iterator it = termfreq.find(tname);
+    if (it != termfreq.end())
+	return it->second;
+    else
+	return 0;
+}
+
+DocumentSet::DocumentSet(const DocumentSet &other)
+    : internal(other.internal)
+{
+}
+
+void
+DocumentSet::operator=(const DocumentSet &other)
+{
+    internal = other.internal;
+}
+
+DocumentSet::DocumentSet()
+    : internal(new Xapian::DocumentSet::Internal)
+{
+}
+
+unsigned int
+DocumentSet::size() const
+{
+    LOGCALL(API, int, "DocumentSet::size", NO_ARGS);
+    return internal->size();
+}
+
+unsigned int
+DocumentSet::Internal::size() const
+{
+    return docs.size();
+}
+
+void
+DocumentSet::add_document(const Document &doc)
+{
+    LOGCALL_VOID(API, "DocumentSet::add_document", doc);
+    internal->add_document(doc);
+}
+void
+DocumentSet::Internal::add_document(const Document &doc)
+{
+    docs.push_back(doc);
+}
+
+Document
+DocumentSet::operator[](doccount i)
+{
+    return internal->get_document(i);
+}
+
+Document
+DocumentSet::Internal::get_document(doccount i)
+{
+    return docs[i];
+}
+
+DocumentSet::~DocumentSet()
+{
+}
+
+class PointTermIterator : public TermIterator::Internal {
+    unordered_map<string, double>::const_iterator i;
+    unordered_map<string, double>::const_iterator end;
+    termcount size;
+    bool started;
+  public:
+    PointTermIterator(const unordered_map<string, double> &termlist) :
+    i(termlist.begin()), end(termlist.end()), size(termlist.size()), started(false)
+    {}
+    termcount get_approx_size() const { return size; }
+    termcount get_wdf() const { throw UnimplementedError("PointIterator doesn't support get_wdf()"); }
+    string get_termname() const { return i->first; }
+    doccount get_termfreq() const { throw UnimplementedError("PointIterator doesn't support get_termfreq()"); }
+    Internal * next();
+    termcount positionlist_count() const {
+	 throw UnimplementedError("PointTermIterator doesn't support positionlist_count()");
+    }
+    bool at_end() const;
+    PositionIterator positionlist_begin() const {
+	throw UnimplementedError("PointTermIterator doesn't support positionlist_begin()");
+    }
+    Internal * skip_to(const string &term);
+};
+
+TermIterator::Internal *
+PointTermIterator::next()
+{
+    if (!started) {
+	started = true;
+	return NULL;
+    }
+    Assert(i != end);
+    ++i;
+    return NULL;
+}
+
+bool
+PointTermIterator::at_end() const
+{
+    if (!started) return false;
+    return i == end;
+}
+
+TermIterator::Internal *
+PointTermIterator::skip_to(const string &term)
+{
+    if (i->first == term)
+	return NULL;
+    while (i->first != term)
```

Comment:
The `if` check here is redundant, since if `i->first == term` we'll drop straight through the loop and `return NULL` at the end of the function.

But actually since the internal container is an `unordered_map` the iteration isn't in term order (it'll be in essentially arbitrary order), so it probably makes more sense to leave `skip_to()` unimplemented.  That's what we do for other instances of `TermIterator` which iterate over non-sorted lists of terms.